src/snowflake/connector/config_parser/parser.py

```diff
@@ -0,0 +1,188 @@
+                f"The value of {self.generate_name()} read from "
+                f"{source} is not part of {self.choices}"
+            )
+        return value
+
+    def generate_name(self) -> str:
+        return ".".join(self._nest_path[1:])
+
+    def generate_env_name(self) -> str:
+        pieces = map(lambda e: e.upper(), self._nest_path[1:])
+        return f"SF{'_' + '_'.join(pieces)}"
+
+    def _get_env(self) -> tuple[bool, str | _T | None]:
+        if self.env_name is False:
+            return False, None
+        if self.env_name is not None:
+            env_name = self.env_name
+        else:
+            # Generate environment name if it wasn't not explicitly supplied,
+            #  and isn't disabled
+            env_name = self.generate_env_name()
+        if env_name not in os.environ:
+            return False, None
+        env_var = os.environ.get(env_name, None)
+        loaded_var: _T | None | str = env_var
+        if env_var and self.type is not None:
+            loaded_var = self.type(env_var)
+        if isinstance(loaded_var, (Table, TOMLDocument)):
+            # If we got a TOML table we probably want it in dictionary form
+            return True, loaded_var.value
+        return True, loaded_var
+
+    def _get_config(self):
+        e = self._root_parser._conf
+        for k in self._nest_path[1:]:
+            e = e[k]
+        if isinstance(e, (Table, TOMLDocument)):
+            # If we got a TOML table we probably want it in dictionary form
+            return e.value
+        return e
+
+
+class ConfigParser:
+    def __init__(
+        self,
+        *,
+        name: str,
+        file_path: Path | None = None,
+    ):
+        self.name = name
+        self.file_path = file_path
+        # Objects holding subparsers and options
+        self._options: dict[str, ConfigOption] = dict()
+        self._sub_parsers: dict[str, ConfigParser] = dict()
+        # Dictionary to cache read in config file
+        self._conf: TOMLDocument | None = None
+        # Information necessary to be able to nest elements
+        #  and add options in O(1)
+        self._root_parser: ConfigParser = self
+        self._nest_path = [name]
+
+    def read_config(
+        self,
+    ) -> None:
+        """Read and parse config file."""
+        if self.file_path is None:
+            raise ConfigParserError(
+                "ConfigParser is trying to read config file, but it doesn't have one"
+            )
+        LOGGER.debug(f"reading configuration file from {str(self.file_path)}")
+        try:
+            self._conf = tomlkit.parse(self.file_path.read_text())
+        except Exception as e:
+            raise ConfigSourceError(
+                f"An unknown error happened while loading '{str(self.file_path)}"
+                f"', please see the error: {e}"
+            )
+
+    @wraps(ConfigOption.__init__)
+    def add_option(
+        self,
+        *args,
+        **kwargs,
```

Comment:
I think we need to document what this `add_option` takes as valid inputs

Reply:
`wraps` does that as `ConfigOption.__init__` has all of the arguments documented. PyCharm (and probably VS code) just don't render these properly